tools/cloudharness-test/cloudharness_test/apitest_init.py

```diff
@@ -66,3 +71,24 @@ def set(self, case, data, context):
                 case.headers = case.headers or {}
                 case.headers["Authorization"] = f"Bearer {data}"
                 case.headers["Cookie"] = f"kc-access={data}"
+
+    UNSAFE_VALUES = ("%")
+
+    @st.hook
+    def before_generate_path_parameters(context: HookContext, strategy):
+        def valid_param(x):
+            # Extract the candidate value.
+            param = x["key"] if isinstance(x, dict) and "key" in x else x
+
+            if param is None or param == "":
+                return True
+
+            param_str = str(param)
+
+            # Reject if any unsafe substring is present.
+            if any(unsafe in param_str for unsafe in UNSAFE_VALUES):
+                return False
+
+            return True
+
+        return strategy.filter(valid_param)
```

Comment:
it also could be done with `filter_path_parameters` which reduces some boilerplate

```python
@st.hook
def filter_path_parameters(context: HookContext, x):
    # Extract the candidate value.
    param = x["key"] if isinstance(x, dict) and "key" in x else x
    
    if param is None or param == "":
        return True

    param_str = str(param)

    # Reject if any unsafe substring is present.
    if any(unsafe in param_str for unsafe in UNSAFE_VALUES):
        return False

    return True

Reply:
Thanks. Could you point me to any docs / code that explain when does the filter_path_parameters happen?

I learned about the before_generate_path_parameters [here](https://schemathesis.readthedocs.io/en/stable/_modules/schemathesis/hooks.html) but I don't see the filter_path_parameters there

Reply:
Yep, the docs are [here](https://schemathesis.readthedocs.io/en/stable/extending.html#hooks) (and a bit more [here](https://schemathesis.readthedocs.io/en/stable/extending.html#filtering-data)). Let me know if they aren't sufficient, but the original idea was to add shortcuts for that nested functions + `strategy.filter` calls

Reply:
It's clear now, thanks. [This ](https://schemathesis.readthedocs.io/en/stable/extending.html#hook-locations) also helped.